cmd/leveldbbench/leveldbbench.go

```diff
@@ -0,0 +1,622 @@
+			Int("bucket", k).
+			Float64("start", start).
+			Float64("end", end).
+			Uint64("count", v).Msg("buckets")
+	}
+	return opCount, buckets
+}
+func writeData(ctx context.Context, db *leveldb.DB, wo *opt.WriteOptions, startIndex, writeLimit uint64, sequential bool) {
+	var i uint64 = startIndex
+	var wg sync.WaitGroup
+	pool := make(chan bool, *degreeOfParallelism)
+	bar := getNewProgressBar(int64(writeLimit), "Writing data")
+	lim := writeLimit + startIndex
+	for ; i < lim; i = i + 1 {
+		pool <- true
+		wg.Add(1)
+		go func(i uint64) {
+			_ = bar.Add(1)
+			k, v := makeKV(i, sizeDistribution.GetSizeSample(), sequential)
+			err := db.Put(k, v, wo)
+			if err != nil {
+				log.Fatal().Err(err).Msg("Failed to put value")
+			}
+			wg.Done()
+			<-pool
+		}(i)
+	}
+	wg.Wait()
+	_ = bar.Finish()
+}
+
+func readSeq(ctx context.Context, db *leveldb.DB, wo *opt.WriteOptions, limit uint64) {
+	pb := getNewProgressBar(int64(limit), "sequential reads")
+	var rCount uint64 = 0
+	pool := make(chan bool, *degreeOfParallelism)
+	var wg sync.WaitGroup
+benchLoop:
+	for {
+		iter := db.NewIterator(nil, nil)
+		for iter.Next() {
+			rCount += 1
+			_ = pb.Add(1)
+			pool <- true
+			wg.Add(1)
+			go func(i iterator.Iterator) {
+				_ = i.Key()
+				_ = i.Value()
+				wg.Done()
+				<-pool
+			}(iter)
+
+			if rCount >= limit {
+				iter.Release()
+				break benchLoop
+			}
+		}
+		iter.Release()
+		err := iter.Error()
+		if err != nil {
+			log.Fatal().Err(err).Msg("Error reading sequentially")
+		}
+	}
+	wg.Wait()
+	_ = pb.Finish()
+}
+func readRandom(ctx context.Context, db *leveldb.DB, ro *opt.ReadOptions, limit uint64) {
+	pb := getNewProgressBar(int64(limit), "random reads")
+	var rCount uint64 = 0
+	pool := make(chan bool, *degreeOfParallelism)
+	var wg sync.WaitGroup
+	rks := NewRandomKeySeeker(db)
+
+benchLoop:
+	for {
+		for {
+			pool <- true
+			wg.Add(1)
+			go func() {
+				rCount += 1
+				_ = pb.Add(1)
+
+				_, err := db.Get(rks.Key(), ro)
+				if err != nil {
+					log.Error().Err(err).Msg("level db random read error")
+				}
+				wg.Done()
+				<-pool
+			}()
+			if rCount >= limit {
+				break benchLoop
+			}
+		}
+	}
+	wg.Wait()
+	_ = pb.Finish()
+}
+
+func NewRandomKeySeeker(db *leveldb.DB) *RandomKeySeeker {
+	rks := new(RandomKeySeeker)
+	rks.db = db
+	rks.iterator = db.NewIterator(nil, nil)
+	rks.firstKey = rks.iterator.Key()
+	return rks
+}
+func (r *RandomKeySeeker) Key() []byte {
+	seekKey := make([]byte, 8)
+	randSrcMutex.Lock()
+	randSrc.Read(seekKey)
+	randSrcMutex.Unlock()
+
+	log.Trace().Str("seekKey", hex.EncodeToString(seekKey)).Msg("searching for key")
+
+	r.iteratorMutex.Lock()
+	defer r.iteratorMutex.Unlock()
+	// first try to just get a random key
+	exists := r.iterator.Seek(seekKey)
+
+	// if that key doesn't exist exactly advance to the next key
+	if !exists {
+		exists = r.iterator.Next()
+	}
+	// if there is no next key, to back to the beginning
+	if !exists {
+		r.iterator.First()
+		r.iterator.Next()
+	}
```

Comment:
curious about how good of a random this is. how often would the initial random key be at a point where there is no keys next in line and having this method return the first key:
```
// first try to just get a random key
exists := r.iterator.Seek(seekKey)
```

Reply:
It largely depends on the number of keys in the underlying level db. If you have 100, then I suspect (I could be wrong) that 1/100 times you'll end up with not greater keys and you just go back to the beginning. 

Where this goes wrong is when the actual keys aren't random. Eg in geth - many of the keys are prefixed with an ascii string like "h". In that case the random won't be as good.

Reply:
I think if we wanted to be more accurate about it, need to take some kind of sample of the actual keys. Maybe a future improvement.